### ISSUE TITLE
Configs are not broadcasted properly in multi-node training

### ISSUE THREAD

**Issue by silverriver** [EleutherAI/gpt-neox#925]
**Describe the bug**
It seems that the `--megatron_config` augment is not properly set for the `train.py` script in multi-node training.

**To Reproduce**

Any script that involve multi-node training setup

**Expected behavior**

The `train.py` script from other nodes (not the main node) should by launched by passing a serizlized json to `--megatron_config`

Something like:

```
[2023-05-05 10:34:36,180] [INFO] [runner.py:550:main] cmd = pdsh -S -f 1024 -w IP1,IP2 python -u -m deepspeed.launcher.launch --world_info=blabla = --node_rank=%n --master_addr=10.234.72.206 --master_port=29500 train.py --deepspeed_config {A serizlized 
 JSON} --megatron_config {A serizlized  JSON}
```

but it seems that the `train.py` script from other nodes (not the main node) is launched by passing a file path to `--megatron_config`

Something like:

```
[2023-05-05 10:34:36,180] [INFO] [runner.py:550:main] cmd = pdsh -S -f 1024 -w IP1,IP2 python -u -m deepspeed.launcher.launch --world_info=blabla = --node_rank=%n --master_addr=10.234.72.206 --master_port=29500 train.py --deepspeed_config {A serizlized 
 JSON} --megatron_config '/data/ubuntu/megatron_config.json'
```

**Proposed solution**
It seems that this issue is caused by this PR #739 

**Screenshots**

**Environment (please complete the following information):**
 - GPUs: 8
- Configs: any configs that use multi nodes

**Additional context**

It seems that the PR #739 breaks this claim for the function `consume_neox_args`:

```
In order not to have any problems with different configs being mismatched across machines, we instead read the .yaml configuration file from the main rank, then serialize the arguments to a dictionary, which the deepspeed launcher broadcasts to all machines (--megatron_config).
```

`--megatron_config` is not broadcasted in the current implementation. Instead, it just pass a local file path to `--megatron_config`. See here:

https://github.com/EleutherAI/gpt-neox/blob/b608043be541602170bfcfb8ec9bf85e8a0799e0/megatron/neox_arguments/arguments.py#L563

**Comment by StellaAthena** [comment]
@silverriver thanks for flagging this. You’re welcome to submit a PR fixing it, or I can try to find some time today to work on it.

**Comment by silverriver** [comment]
It will be great if you can make the PR. I will also look into the detailed implementation later and try to figure out how to fix it.

**Comment by silverriver** [comment]
BTW, I am also wondering why would we use `base64.urlsafe_b64decode` to encode the config json. Why not just use the serizlized json str produced by `json.dumps` ?

https://github.com/EleutherAI/gpt-neox/blob/b608043be541602170bfcfb8ec9bf85e8a0799e0/megatron/neox_arguments/arguments.py#L435

**Comment by StellaAthena** [comment]
> BTW, I am also wondering why would we use `base64.urlsafe_b64decode` to encode the config json. Why not just use the serizlized json str produced by `json.dumps` ?
> 
> https://github.com/EleutherAI/gpt-neox/blob/b608043be541602170bfcfb8ec9bf85e8a0799e0/megatron/neox_arguments/arguments.py#L435

@dashstander it looks like you introduced this.

**Comment by dashstander** [comment]
There are a number of reasons this was introduced that all basically revolve around getting DeepSpeed autotuning to work with Slurm. In particular the pattern of doing the base 64 encoding is a[ pattern borrowed from DeepSpeed](https://github.com/microsoft/DeepSpeed/blob/4e886f0568832d292183926bcc1a9105def25f2c/deepspeed/autotuning/scheduler.py#L337). Though this did have the added benefit of keeping slurm's argument parser happy--it really didn't like the proliferation of double quotes in your average json string.

I'm sorry that this isn't working for you, though I think we will need a little bit more information on your setup to reproduce. It occurs to me that all of the development and testing I did for this occurred on clusters where the configs (the initial ones that you give to `deepy.py`) are all saved in a shared filesystem that all of the worker nodes can access. Is that where this is going wrong with your cluster?

**Comment by silverriver** [comment]
Thank for the clarification.

In my case, I put (duplicated) code and config files to the local disk of each worker node.
Not all nodes can get access to `megatron_config` since it is only produced by the rank0-node. I think this setting leads to the error.

Although we have a shared file system. But we only use it for hosting training data and saving ckpt for now.
Maybe my setting is not the best way to launch multi-node training. We will try to put all codes and configs files on the shared file system so that all nodes can get access to the same set of files. 

But either way I think we should consider to revised the doc string of the function `consume_neox_args` and maybe add more docs about this behavior (i.e., to enable multi-node training, we need to host each nodes on a shared file system so that all nodes can get access to the same code folder). 

Moreover, for users that do not have a shared file system. We still need to find a way to pass the config to all nodes when launching the multi-node training.

**Comment by dashstander** [comment]
Oh absolutely we should figure out how to make this work if you don't have access to a shared filesystem. I would just want to make sure that whatever changes we make don't break autotuning

**Comment by StellaAthena** [comment]
@dashstander I have reproduced this on our CoreWeave cluster. I'm having trouble figuring out why it worked before though... it doesn't seem like it should. However I have verified that it runs using this commit: https://github.com/EleutherAI/gpt-neox/commit/68d223cab68aff0c9c5b74da3772503b8b3a2acd

**Comment by dashstander** [comment]
Ok, since I'm the most familiar with the autotuning code let me take a crack at fixing this.

**Comment by dashstander** [comment]
@silverriver does PR #948 fix your issue?

**Comment by silverriver** [comment]
I think the PR fixes the config issue.

Thank you very much
